mqtt/mqtt-broker/benches/dispatch_messages.rs

```diff
@@ -0,0 +1,444 @@
+use std::{
+    collections::HashSet,
+    iter::FromIterator,
+    sync::{Arc, Mutex},
+    time::Duration,
+};
+
+use bytes::Bytes;
+use criterion::{
+    criterion_group, criterion_main, measurement::WallTime, BatchSize, BenchmarkGroup, Criterion,
+};
+use itertools::iproduct;
+use rand::Rng;
+use tokio::{
+    runtime::Runtime,
+    sync::mpsc::{self, UnboundedReceiver},
+};
+use tracing::{info, warn};
+
+use mqtt3::{proto, PROTOCOL_LEVEL, PROTOCOL_NAME};
+use mqtt_broker::{
+    AuthId, BrokerBuilder, BrokerHandle, ClientEvent, ClientId, ConnReq, ConnectionHandle, Message,
+    Publish, SystemEvent,
+};
+
+criterion_group!(basic, one_to_one, fan_in, fan_out);
+criterion_main!(basic);
+
+fn one_to_one(c: &mut Criterion) {
+    init_logging();
+
+    let mut group = c.benchmark_group("1-to-1");
+    for (qos, count, payload_size) in scenarios() {
+        let strategy = Strategy::new_one_to_one(count, count);
+        dispatch_messages(&mut group, strategy, count, payload_size, qos);
+    }
+    group.finish();
+}
+
+fn fan_in(c: &mut Criterion) {
+    init_logging();
+
+    let mut group = c.benchmark_group("fan_in");
+    for (qos, count, payload_size) in scenarios() {
+        let strategy = Strategy::new_fan_in(count, count);
+        dispatch_messages(&mut group, strategy, count, payload_size, qos);
+    }
+    group.finish();
+}
+
+fn fan_out(c: &mut Criterion) {
+    init_logging();
+
+    let mut group = c.benchmark_group("fan_out");
+    for (qos, count, payload_size) in scenarios() {
+        let strategy = Strategy::new_fan_out(count, count);
+        dispatch_messages(&mut group, strategy, count, payload_size, qos);
+    }
+    group.finish();
+}
+
```

Comment:
these three methods look exactly the same, except the strategy they create and the name. Can it be made into a single method, and fan_in()/fan_out()/one_to_one() just call that single method with a "strategy creator" function injected?

Reply:
Yeah, I thought about that. I still wonder if number of publishers should not be equal number of subscribers. it may vary depending on strategy...